mlflow/pyfunc/__init__.py

```diff
@@ -1502,151 +1503,148 @@ def udf(
 
         # Note: this is a pandas udf function in iteration style, which takes an iterator of
         # tuple of pandas.Series and outputs an iterator of pandas.Series.
-        # Save original environment variables so that we convert it back after udf task
-        original_envs = {}
+        update_envs = {}
         if mlflow_home is not None:
-            original_envs["MLFLOW_HOME"] = os.environ.get("MLFLOW_HOME")
-            os.environ["MLFLOW_HOME"] = mlflow_home
+            update_envs["MLFLOW_HOME"] = mlflow_home
         if openai_env_vars:
-            original_envs.update({k: os.environ.get(k) for k in openai_env_vars})
-            os.environ.update(openai_env_vars)
+            update_envs.update(openai_env_vars)
         if mlflow_testing:
-            original_envs[_MLFLOW_TESTING.name] = os.environ.get(_MLFLOW_TESTING.name)
-            _MLFLOW_TESTING.set(mlflow_testing)
-        if extra_env_vars:
-            original_envs.update({k: os.environ.get(k) for k in extra_env_vars})
-            os.environ.update(extra_env_vars)
+            update_envs[_MLFLOW_TESTING.name] = mlflow_testing
+        if extra_env:
+            update_envs.update(extra_env)
+
+        #  use `modified_environ` to temporarily set the envs and restore them finally
+        with modified_environ(update=update_envs):
```

Comment:
nit:

can we make `modified_environ` to be a decorator, like:
```
    @pandas_udf(result_type)
    @modified_environ(update={...})  # decorate here
    def udf(
        iterator: Iterator[Tuple[Union[pandas.Series, pandas.DataFrame], ...]]
    ) -> Iterator[result_type_hint]:
  ...
```

so that we don't need to change so many lines in this PR, and we can avoid so many lines indent.

Reply:
what's wrong with changing so many lines?

Reply:
I think decorator doesn't work here, as we need to pass the environment variables into the udf, if using decorator it's still changing outside of the udf.

Reply:
I think a simple decorator like this should work (I haven't tested though)

```
def modified_environ_decorator(update):
    
    def dec(f):
        def decorated_f(*args, **kwargs):
            with modified_environ(update=update):
                return f(*args, **kwargs)
    
    return dec

    @pandas_udf(result_type)
    @modified_environ_decorator(update={...})  # decorate here
    def udf(
        iterator: Iterator[Tuple[Union[pandas.Series, pandas.DataFrame], ...]]
    ) -> Iterator[result_type_hint]:
```

Reply:
It doesn't work, I tested. Because we need to update envs inside `f(*args, **kwargs)`

Reply:
Ok now I tested in https://github.com/mlflow/mlflow/pull/10605/commits/f9b177ecb00ce032f907ee7d4258264d14a99194 and the test fails, so I think it doesn't work...

Reply:
Thanks, I will check, but this is not a blocker.